lib/core/src/test/java/io/atlasmap/actions/ExpressionFieldActionTest.java

```diff
@@ -189,4 +189,31 @@ public void testNumberField() throws Exception {
         assertEquals("bigdecimal 1", ExpressionFieldAction.process(action, Arrays.asList(bigDec)));
     }
 
+    @Test
+    public void testConcatenateActionWithDelimiter() throws Exception {
+        Expression action = new Expression();
+        action.setExpression("CONCATENATE{delimiter=${0}}(${1}, ${2}, ${3})");
+        assertEquals("a,b,c", ExpressionFieldAction.process(action, Arrays.asList(",", "a", "b", "c")));
+    }
+
+    @Test
+    public void testConcatenateActionWithWrongProperty() throws Exception {
+        Expression action = new Expression();
+        action.setExpression("CONCATENATE{delimiter=${0},fail='true'}(${1}, ${2}, ${3})");
```

Comment:
Please note the format I picked for providing action parameters...

Reply:
Wow, it's great job that you found a way to assemble this into current code.
I was thinking about squeezing those properties as function arguments. And I understood that is not possible because the action detail is not unique with just a name, but it has to also look at the value type (for overloading).
What do you think, instead, separating field action id from display name so that field action ID could be always unique? If you can find a unique action detail with an ID, then you can pass function arguments to `ActionParameters` as much as it's declared, then we can assume that rest are the values to process. Once we get syntax assist in the UI expression input, then we can provide `ActionParameters` metadata as a part of argument description. We might also have to take care of the order of `ActionParameters`.
I think we can defer [1] as I don't think there're only a few if there is which has conflicting name. I hope the order [2] is always identical on initialization, but have to investigate.
1. Make field action id unique and separate from name
1. Take care of `ActionParameters` order for function usage

Reply:
Regarding [1] I'm not sure, if I follow you. In the current implementation I do take care of identifying the value type at https://github.com/atlasmap/atlasmap/pull/1610/files#r359283060 for the purpose of selecting the right action if there are multiple actions with the same name. 

Regarding the format [2]. Do we want to make it CONCATENATE(delimiter=${0}, fail='true', ${1}, ${2}, ${3}) i.e. use named arguments as in e.g. Python? I wouldn't simply assume that first arguments are for ActionParameters as they may be optional e.g. delimiter can be omitted, because it's "" by default.

Reply:
I'd like to also ask @chirino's opinion here

Ah named arguments sound better IMO. But still I wonder which of this named arguments and disallow optional property is better - if we can auto-fill default property values in the UI as a part of expression syntax assist, I think we can make all action properties mandatory, thus just the order of arguments can identify them.

Reply:
Okay, let's wait for @chirino to chime in. I've fixed the code to use named arguments.

Reply:
As a side note, thinking about the syntax assist as well as a tooltip we would show for each function in UI, relying on the order of arguments would make it simpler.

Reply:
Just to clarify do you think that expecting action parameters as first arguments i.e. CONCATENATE(${0}, ${1}, ${2}, ${3}), where ${0} is a delimiter, is simpler than using the named arguments i.e. CONCATENATE(delimiter=${0}, ${1}, ${2}, ${3})?

Do we have the syntax assist in place already? I see us being able to support both ways at some point (just as Python does) and think that named arguments are more readable until the syntax assist is in place. Also relying on the order becomes less readable when more arguments are expected and some with defaults, but I agree that with the syntax assist it is equally usable.

Reply:
Yes, actually some of OOTB action has more than one action parameters, but I assumed that is already identical, as `ActionDetail` holds `ActionParameters`. Is that true? So if order is fixed, runtime should be able to internally identify which is action parameter(s) and the rest are field values.
IMO named argument is still a bit tricky for citizen integrator=non-programmer, and that somehow breaks the sales copy "Spreadsheet like simple math formula". Then when we want to show a tooltip help for the functions, if argument order could be changed with named argument, that is a bit more confusing.
Unfortunately we don't have syntax assist yet, actually after introducing dynamic path expression (#1512, field ref is changed to like ${docId:/path/to/field}) we'd have to rebuild expression UI.
So as far as argument order is fixed, introducing named arguments for the temporary solution until we deliver syntax support, could be an option I think. We can keep named argument support, but just let UI to fill those without name.

Reply:
Just noticed that, even before we support syntax assist, we can describe those functions in manual. And I think fixed order arguments makes it a bit easier to read, as user don't have to find the argument name in the manual and type it in UI.

Reply:
Okay, let me do the fixed order arguments. I've noticed that argument names are not consistently displayed in the current UI under transformations so indeed one would have to look in a manual to figure them out.